lib/SILGen/SILGenApply.cpp

```diff
@@ -6067,6 +6067,39 @@ RValue SILGenFunction::emitDynamicSubscriptExpr(DynamicSubscriptExpr *e,
   return RValue(*this, e, emitManagedRValueWithCleanup(optResult, optTL));
 }
 
+SmallVector<ManagedValue, 4> SILGenFunction::emitKeyPathSubscriptOperands(
+    SubscriptDecl *subscript, SubstitutionMap subs, Expr *indexExpr) {
+  Type interfaceType = subscript->getInterfaceType();
+  CanFunctionType substFnType =
+      subs ? cast<FunctionType>(interfaceType->castTo<GenericFunctionType>()
+                                    ->substGenericArgs(subs)
+                                    ->getCanonicalType())
+           : cast<FunctionType>(interfaceType->getCanonicalType());
+  AbstractionPattern origFnType(substFnType);
+  auto fnType =
+      getLoweredType(origFnType, substFnType).castTo<SILFunctionType>();
+
+  SmallVector<ManagedValue, 4> argValues;
+  SmallVector<DelayedArgument, 2> delayedArgs;
+  ArgEmitter emitter(*this, fnType->getRepresentation(),
```

Comment:
Nice! Using ArgEmitter is definitely the right approach here.

It appears that the old code didn't correctly handle keypaths with variadic subscript components either, eg

```
struct S {
  subscript<T>(x: T...) -> Int { get {} set {} }
}

let k = \S.[1, 2, 3]
```

Crashes with this on master

```
Assertion failed: (hashables.size() == SubscriptSize), function setSubscriptIndexHashableConformances, file /Users/spestov/src/swift/lib/AST/Expr.cpp, line 2129.
Stack dump:
0.	Program arguments: /Users/spestov/src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/bin/swift -frontend -c -primary-file var.swift -target x86_64-apple-darwin19.0.0 -enable-objc-interop -color-diagnostics -module-name var -o /var/folders/wg/nmvf884n4b14lt0yw4khjsv00000gn/T/var-9e61ce.o 
1.	Swift version 5.1.1-dev (Swift 69c2de1727)
2.	While type-checking statement at [var.swift:5:1 - line:5:20] RangeText="let k = \S.[1, 2, 3"
3.	While type-checking declaration 0x7fd9ed0ca3a0 (at var.swift:5:1)
4.	While evaluating request PatternBindingEntryRequest((unknown decl), 0)
5.	While type-checking expression at [var.swift:5:9 - line:5:20] RangeText="\S.[1, 2, 3"
```

It looks like a similar issue. Can you see if your fix addresses this as well? If not, a follow up PR would be great :)

Reply:
I'll have to create a follow-up patch for variadics as it looks like that will require some additional work.